core/src/main/java/io/micronaut/core/execution/DelayedExecutionFlowImpl.java

```diff
@@ -0,0 +1,449 @@
+        return next(new OnErrorResume(fallback));
+    }
+
+    @Override
+    public ExecutionFlow<T> putInContext(String key, Object value) {
+        return this;
+    }
+
+    @Override
+    public void onComplete(BiConsumer<? super T, Throwable> fn) {
+        next(new OnComplete<>(fn));
+    }
+
+    @SuppressWarnings("unchecked")
+    @Nullable
+    @Override
+    public ImperativeExecutionFlow<T> tryComplete() {
+        Object tailOutput = tail.output;
+        if (tailOutput != null) {
+            if (tailOutput instanceof Failure failure) {
+                return (ImperativeExecutionFlow<T>) new ImperativeExecutionFlowImpl(null, failure.t);
+            } else if (tailOutput == NULL) {
+                return (ImperativeExecutionFlow<T>) new ImperativeExecutionFlowImpl(null, null);
+            } else {
+                return (ImperativeExecutionFlow<T>) new ImperativeExecutionFlowImpl(tailOutput, null);
+            }
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Special wrapper for exception results.
+     *
+     * @param t The exception of the failure
+     */
+    private record Failure(Throwable t) {
+    }
+
+    private abstract static class Step {
+        /**
+         * The next step to take, or {@code null} if there is no next step yet.
+         */
+        private volatile Step next;
+        /**
+         * The output of this step, or {@code null} if this step has not completed yet.
+         */
+        private volatile Object output;
+
+        /**
+         * Apply this step. Must call one of {@link #returnImmediate}, {@link #returnFlow},
+         * {@link #returnError} or {@link #returnUnchanged}.
+         *
+         * @param input The input for the step
+         * @return The return value of the {@code return*} method called
+         */
+        abstract Object apply(Object input);
+
+        /**
+         * Atomically set the output of this step. If this returns non-null, the caller must call
+         * {@link #work(Step, Object)} with the returned step.
+         *
+         * @param output The output of this step
+         * @return The next step to execute using {@link #work(Step, Object)}, or {@code null} if
+         * the next step will be executed later
+         */
+        @Nullable
+        final Step atomicSetOutput(Object output) {
+            if (this.output != null) {
+                // this is a best-effort check, the output field isn't always set
+                throw new IllegalStateException("Already completed");
+            }
+            Step next = this.next;
+            if (next != null) {
+                return next;
+            }
+            this.output = output;
+            next = this.next;
+            if (next != null) {
+                // another thread completed at the same time! one or both threads hit this sync
+                // block.
+                synchronized (this) {
```

Comment:
it could be implemented using a CAS but i think this is simpler and it should almost never be hit anyway. for loom it should be fine since it's only blocked for a very short time